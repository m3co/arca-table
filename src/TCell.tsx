import * as React from 'react';

import { Input } from './Input';
import { Search } from './Search';

import { Row, Field, Fields, SearchMethod } from 'arca-redux/';

import './TCell.less';
import { bool } from 'prop-types';

interface Props {
  Field: Field;
  Row: Row;
  onEdit?: (Row: Row, column?: keyof Fields, Field?: Field) => void;
  search?: SearchMethod;
}

interface State {
  edit: boolean;
  dirty: boolean;
  value: string | number | boolean;
  column?: keyof Fields;
}

export class TCell
  extends React.Component<Props, State>
{
  public state: State = {
    edit: false,
    dirty: false,
    value: '',
  }

  public constructor(props: Props) {
    super(props);

    const { Field, Row } = this.props;
    const columns = Object.keys(Row) as (keyof Fields)[];
    const column = columns.find((column): boolean =>
      column === Field.Name
    );
    if (column) {
      const row = Row as Fields;
      const value = row[column];
      if (Field.Type.toLowerCase() == 'checkbox') {
        this.state.value = Boolean(value);
      } else {
        this.state.value = value ? value.toString() : '';
      }
      this.state.column = column;
    }
  }

  public componentDidUpdate(prevProps: Props): void {
    const { props, props: { Field: { Type } }, state } = this;
    if (state.column && prevProps.Row !== props.Row) {
      const row = props.Row as Fields;
      const value = row[state.column];
      this.setState((state: State): State => {
        return {
          ...state,
          value: Type.toLowerCase() == 'checkbox' ?
            Boolean(value) : (value ? value.toString() : ''),
          dirty: false
        };
      });
    }
  }

  private startEdit = (): void => {
    if (this.props.Field.Primary) return;
    if (!this.props.Field.Editable) return;
    this.setState((state: State): State => {
      return {
        ...state,
        edit: true,
      };
    });
  }

  private cancelEdit = (): void => {
    this.setState((state: State): State => {
      return {
        ...state,
        edit: false,
      };
    });
  }

  private finishEdit = (column: keyof Fields):
    (currentValue: string | number | boolean) => void =>
    (currentValue: string | number | boolean): void => {
    this.setState((state: State): State => {
      if (this.props.onEdit) {
        const { Field, Row } = this.props;
        const newRow = {
          ...Row,
          [column]: currentValue,
        };
        (currentValue !== state.value) && this.props.onEdit(newRow, column, Field);
      }
      return {
        ...state,
        value: currentValue,
        edit: false,
        dirty: currentValue !== state.value,
      }
    });
  }

  private onChangeCheckbox = (column: keyof Fields): () => void => {
    return (): void => {
      const { Field, Row } = this.props;
      if (!Field.Editable) return;
      this.setState((state: State): State => {
        const currentValue = !state.value;
        if (this.props.onEdit) {
          const newRow = {
            ...Row,
            [column]: currentValue,
          };
          this.props.onEdit(newRow, column, Field);
        }
        return {
          ...state,
          value: currentValue,
          edit: false,
          dirty: true,
        }
      })
    }
  }

  public render(): JSX.Element {
    const { Field, Row, search } = this.props;
    const row = Row as Fields;
    const { Combobox } = Field;
    const { value, column, dirty } = this.state;
    const className = `${Field.Primary ? 'primary' : ''} ${dirty ? 'dirty' : ''}`.trim();
    if (column && Field.Type.toLowerCase() == 'checkbox') {
      return (
        <td>
          <input
            type={Field.Type}
            checked={!!value}
            onChange={this.onChangeCheckbox(column)} />
        </td>
      );
    }
    return (column) ?
      (
        <td
          className={className ? className : undefined}
          onClick={this.startEdit}>
          {(this.state.edit) ? (
            (search && Combobox) ? (
              <Search
                search={search}
                Row={Row}
                Setup={Combobox}
                value={value.toString()}
                onBlur={this.finishEdit(column)}
                onEnter={this.finishEdit(column)}
                onEsc={this.cancelEdit} />
            ) : (
              <Input
                type={Field.Type}
                value={value.toString()}
                onBlur={this.finishEdit(column)}
                onEnter={this.finishEdit(column)}
                onEsc={this.cancelEdit}
              />
            )
          ) : (dirty) ? value : row[column]}
        </td>
      ) :
      (<td>-</td>);
  }
}
