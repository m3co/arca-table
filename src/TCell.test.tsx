
import * as React from 'react';
import { mount } from 'enzyme';

import { TCell } from './TCell';
import { State, Params, FACADReports, Field, Fields, Row } from 'arca-redux';

interface Response {
  ID: string;
  Method: 'Search';
  Context: {
    Field: keyof Fields;
    Source: keyof State['Source'];
  };
  Result: Params[];
  Error: {} | null;
};

const searchMock = (Source: keyof State['Source'],
  Field: keyof Fields,
  Params: Params): Promise<Response> => {
  const response: Response = {
    ID: '',
    Method: 'Search',
    Context: {
      Field,
      Source,
    },
    Result: [
      {BuiltInCategory: 'OST_InternalLineLoads'},
      {BuiltInCategory: 'OST_InternalLineLoadTags'},
      {BuiltInCategory: 'OST_LineLoads'},
      {BuiltInCategory: 'OST_LineLoadTags'},
    ],
    Error: null,
  };
  return Promise.resolve(response);
};

interface MockHanders {
  onEdit?: (Row: Row, column?: keyof Fields, Field?: Field) => void;
  search?: typeof searchMock;
  isNew?: boolean;
};

function prepareMock1(overloadField?: Field, overloadedHandlers?: MockHanders): JSX.Element {
  const row: FACADReports['Row'] = {
    ID: 1,
    BuiltInCategory: 'INVALID',
    ReportType: 'Schedule',
    Name: 'A name',
    PathName: 'A path',
    Field1: '',
    Field2: '',
    Field3: '',
  };

  const field: Field = overloadField || {
    Editable: true,
    Primary: false,
    Name: 'Name',
    Type: 'Text',
    Required: false,
  };

  const handlers: MockHanders = overloadedHandlers || {};
  return (
    <table>
      <tbody>
        <tr>
          <TCell Row={row} Field={field} onEdit={handlers.onEdit} search={handlers.search} isNew={handlers.isNew} />
        </tr>
      </tbody>
    </table>
  );
}

test('Cell renders a normal cell', (): void => {
  const el = prepareMock1();
  const wrapper = mount(el);
  expect(wrapper).toMatchSnapshot();
});

test('Cell renders a combobox cell', (): void => {
  const field: Field = {
    Name: 'BuiltInCategory',
    Type: 'Text',
    Primary: false,
    Required: true,
    Editable: true,
    Combobox:{
      Source: 'FACAD-BuiltInCategories',
      Display: 'BuiltInCategory',
      Value: 'BuiltInCategory',
      Params:{
        BuiltInCategory: 'BuiltInCategory'
      }
    }
  };
  const el = prepareMock1(field, { search: searchMock });
  const wrapper = mount(el);
  wrapper.find('TCell').simulate('click');
  expect(wrapper.find('TCell Search datalist').length).toBe(1);
});

test('Cell renders an empty cell if field is incorrect', (): void => {
  const field: Field = {
    Editable: true,
    Primary: false,
    Name: 'Name Incorrect',
    Type: 'Text',
    Required: false,
  }

  const el = prepareMock1(field);
  const wrapper = mount(el);
  expect(wrapper).toMatchSnapshot();
});

test('Click over cell renders an input', (): void => {
  const el = prepareMock1();
  const wrapper = mount(el);
  wrapper.find('TCell').simulate('click');
  expect(wrapper.find('TCell input').length).toBe(1);
  expect(wrapper).toMatchSnapshot();
});

test('Click over a primary cell editable - edit', (): void => {
  const field: Field = {
    Editable: true,
    Primary: true,
    Name: 'ID',
    Type: 'Text',
    Required: false,
  }

  const el = prepareMock1(field);
  const wrapper = mount(el);
  wrapper.find('TCell').simulate('click');
  expect(wrapper.find('TCell input').length).toBe(1);
  expect(wrapper).toMatchSnapshot();
});

test('Click over a new and primary cell non-editable but required - edit', (): void => {
  const field: Field = {
    Editable: false,
    Primary: true,
    Name: 'ID',
    Type: 'Text',
    Required: true,
  }

  const el = prepareMock1(field, {isNew: true});
  const wrapper = mount(el);
  wrapper.find('TCell').simulate('click');
  expect(wrapper.find('TCell input').length).toBe(1);
  expect(wrapper).toMatchSnapshot();
});

test('Click over a cell non-editable - nothing', (): void => {
  const field: Field = {
    Editable: false,
    Primary: false,
    Name: 'BuiltInCategory',
    Type: 'Text',
    Required: false,
  }

  const el = prepareMock1(field);
  const wrapper = mount(el);
  wrapper.find('TCell').simulate('click');
  expect(wrapper.find('TCell input').length).toBe(0);
  expect(wrapper).toMatchSnapshot();
});

test('Cancel edit by pressing ESC', (): void => {
  const field: Field = {
    Editable: true,
    Primary: false,
    Name: 'BuiltInCategory',
    Type: 'Text',
    Required: false,
  }

  const el = prepareMock1(field);
  const wrapper = mount(el);
  wrapper.find('TCell').simulate('click');
  expect(wrapper.find('TCell input').length).toBe(1);
  wrapper.find('input').simulate('keyup', { key: 'Escape' });
  expect(wrapper.find('TCell input').length).toBe(0);
});

test('Finish edit if blur', (): void => {
  const newValue = 'modified value';
  const field: Field = {
    Editable: true,
    Primary: false,
    Name: 'BuiltInCategory',
    Type: 'Text',
    Required: false,
  }

  const el = prepareMock1(field);
  const wrapper = mount(el);
  wrapper.find('TCell').simulate('click');
  expect(wrapper.find('TCell input').length).toBe(1);
  wrapper.find('input').simulate('change', { target: { value: newValue } });
  wrapper.find('input').simulate('blur');
  expect(wrapper.find('TCell').text()).toBe(newValue);
  expect(wrapper).toMatchSnapshot();
});

test('Finish edit if Enter', (): void => {
  const newValue = 'modified value';
  const field: Field = {
    Editable: true,
    Primary: false,
    Name: 'BuiltInCategory',
    Type: 'Text',
    Required: false,
  }

  const el = prepareMock1(field);
  const wrapper = mount(el);
  wrapper.find('TCell').simulate('click');
  expect(wrapper.find('TCell input').length).toBe(1);
  wrapper.find('input').simulate('change', { target: { value: newValue } });
  wrapper.find('input').simulate('keyup', { key: 'Enter' });
  expect(wrapper.find('TCell').text()).toBe(newValue);
  expect(wrapper).toMatchSnapshot();
});

test('Finish edit fires onEdit', (): void => {
  const newValue = 'modified value';
  const field: Field = {
    Editable: true,
    Primary: false,
    Name: 'BuiltInCategory',
    Type: 'Text',
    Required: false,
  }

  let onEditReached = false;
  const handlers: MockHanders = {
    onEdit: (Row: Row, column?: keyof Fields, Field?: Field): void => {
      const row = Row as Fields;
      if (column) {
        expect(row[column]).toBe(newValue);
      } else {
        fail();
      }
      onEditReached = true;
    }
  }

  const el = prepareMock1(field, handlers);
  const wrapper = mount(el);
  wrapper.find('TCell').simulate('click');
  expect(wrapper.find('TCell input').length).toBe(1);
  wrapper.find('input').simulate('change', { target: { value: newValue } });
  wrapper.find('input').simulate('keyup', { key: 'Enter' });
  expect(wrapper.find('TCell').text()).toBe(newValue);
  expect(wrapper).toMatchSnapshot();
  expect(onEditReached).toBe(true);
});
