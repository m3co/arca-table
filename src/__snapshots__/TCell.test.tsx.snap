// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell renders a normal cell 1`] = `
<table>
  <tbody>
    <tr>
      <TCell
        Field={
          Object {
            "Editable": true,
            "Name": "Name",
            "Primary": false,
            "Required": false,
            "Type": "Text",
          }
        }
        Row={
          Object {
            "BuiltInCategory": "INVALID",
            "Field1": "",
            "Field2": "",
            "Field3": "",
            "ID": 1,
            "Name": "A name",
            "PathName": "A path",
            "ReportType": "Schedule",
          }
        }
      >
        <td
          onClick={[Function]}
        >
          A name
        </td>
      </TCell>
    </tr>
  </tbody>
</table>
`;

exports[`Cell renders an empty cell if field is incorrect 1`] = `
<table>
  <tbody>
    <tr>
      <TCell
        Field={
          Object {
            "Editable": true,
            "Name": "Name Incorrect",
            "Primary": false,
            "Required": false,
            "Type": "Text",
          }
        }
        Row={
          Object {
            "BuiltInCategory": "INVALID",
            "Field1": "",
            "Field2": "",
            "Field3": "",
            "ID": 1,
            "Name": "A name",
            "PathName": "A path",
            "ReportType": "Schedule",
          }
        }
      >
        <td>
          -
        </td>
      </TCell>
    </tr>
  </tbody>
</table>
`;

exports[`Click over a cell non-editable - nothing 1`] = `
<table>
  <tbody>
    <tr>
      <TCell
        Field={
          Object {
            "Editable": false,
            "Name": "BuiltInCategory",
            "Primary": false,
            "Required": false,
            "Type": "Text",
          }
        }
        Row={
          Object {
            "BuiltInCategory": "INVALID",
            "Field1": "",
            "Field2": "",
            "Field3": "",
            "ID": 1,
            "Name": "A name",
            "PathName": "A path",
            "ReportType": "Schedule",
          }
        }
      >
        <td
          onClick={[Function]}
        >
          INVALID
        </td>
      </TCell>
    </tr>
  </tbody>
</table>
`;

exports[`Click over a primary cell - nothing 1`] = `
<table>
  <tbody>
    <tr>
      <TCell
        Field={
          Object {
            "Editable": true,
            "Name": "ID",
            "Primary": true,
            "Required": false,
            "Type": "Text",
          }
        }
        Row={
          Object {
            "BuiltInCategory": "INVALID",
            "Field1": "",
            "Field2": "",
            "Field3": "",
            "ID": 1,
            "Name": "A name",
            "PathName": "A path",
            "ReportType": "Schedule",
          }
        }
      >
        <td
          className="primary"
          onClick={[Function]}
        >
          <Input
            onBlur={[Function]}
            onEnter={[Function]}
            onEsc={[Function]}
            type="Text"
            value="1"
          >
            <input
              autoFocus={true}
              onBlur={[Function]}
              onChange={[Function]}
              onKeyUp={[Function]}
              type="Text"
              value="1"
            />
          </Input>
        </td>
      </TCell>
    </tr>
  </tbody>
</table>
`;

exports[`Click over cell renders an input 1`] = `
<table>
  <tbody>
    <tr>
      <TCell
        Field={
          Object {
            "Editable": true,
            "Name": "Name",
            "Primary": false,
            "Required": false,
            "Type": "Text",
          }
        }
        Row={
          Object {
            "BuiltInCategory": "INVALID",
            "Field1": "",
            "Field2": "",
            "Field3": "",
            "ID": 1,
            "Name": "A name",
            "PathName": "A path",
            "ReportType": "Schedule",
          }
        }
      >
        <td
          onClick={[Function]}
        >
          <Input
            onBlur={[Function]}
            onEnter={[Function]}
            onEsc={[Function]}
            type="Text"
            value="A name"
          >
            <input
              autoFocus={true}
              onBlur={[Function]}
              onChange={[Function]}
              onKeyUp={[Function]}
              type="Text"
              value="A name"
            />
          </Input>
        </td>
      </TCell>
    </tr>
  </tbody>
</table>
`;

exports[`Finish edit fires onEdit 1`] = `
<table>
  <tbody>
    <tr>
      <TCell
        Field={
          Object {
            "Editable": true,
            "Name": "BuiltInCategory",
            "Primary": false,
            "Required": false,
            "Type": "Text",
          }
        }
        Row={
          Object {
            "BuiltInCategory": "INVALID",
            "Field1": "",
            "Field2": "",
            "Field3": "",
            "ID": 1,
            "Name": "A name",
            "PathName": "A path",
            "ReportType": "Schedule",
          }
        }
        onEdit={[Function]}
      >
        <td
          className="dirty"
          onClick={[Function]}
        >
          modified value
        </td>
      </TCell>
    </tr>
  </tbody>
</table>
`;

exports[`Finish edit if Enter 1`] = `
<table>
  <tbody>
    <tr>
      <TCell
        Field={
          Object {
            "Editable": true,
            "Name": "BuiltInCategory",
            "Primary": false,
            "Required": false,
            "Type": "Text",
          }
        }
        Row={
          Object {
            "BuiltInCategory": "INVALID",
            "Field1": "",
            "Field2": "",
            "Field3": "",
            "ID": 1,
            "Name": "A name",
            "PathName": "A path",
            "ReportType": "Schedule",
          }
        }
      >
        <td
          className="dirty"
          onClick={[Function]}
        >
          modified value
        </td>
      </TCell>
    </tr>
  </tbody>
</table>
`;

exports[`Finish edit if blur 1`] = `
<table>
  <tbody>
    <tr>
      <TCell
        Field={
          Object {
            "Editable": true,
            "Name": "BuiltInCategory",
            "Primary": false,
            "Required": false,
            "Type": "Text",
          }
        }
        Row={
          Object {
            "BuiltInCategory": "INVALID",
            "Field1": "",
            "Field2": "",
            "Field3": "",
            "ID": 1,
            "Name": "A name",
            "PathName": "A path",
            "ReportType": "Schedule",
          }
        }
      >
        <td
          className="dirty"
          onClick={[Function]}
        >
          modified value
        </td>
      </TCell>
    </tr>
  </tbody>
</table>
`;
